base/broadcast.jl

```diff
@@ -3,63 +3,207 @@
 module Broadcast
 
 using Base.Cartesian
-using Base: linearindices, tail, OneTo, to_shape,
+using Base: Bottom, Indices, OneTo, linearindices, tail, to_shape,
             _msk_end, unsafe_bitgetindex, bitcache_chunks, bitcache_size, dumpbitcache,
             nullable_returntype, null_safe_op, hasvalue, isoperator
 import Base: broadcast, broadcast!
 export broadcast_getindex, broadcast_setindex!, dotview, @__dot__
 
-const ScalarType = Union{Type{Any}, Type{Nullable}}
+# Note: `indices` will be overridden below, thus you need to use
+# Base.indices when you want the Base versions.
+
+## Types used by `rule`
+# Unknown acts a bit like `Bottom`, in that it loses to everything. But by not having
+# it be a subtype of every other type, we limit the need for ambiguity resolution.
+abstract type Unknown end
+# Objects that act like a scalar for purposes of broadcasting
+abstract type Scalar end
+# An AbstractArray type that "loses" in precedence comparisons to all other AbstractArrays.
+# We will want to keep track of dimensionality, so we make it the first parameter.
+abstract type BottomArray{N} <: AbstractArray{Void,N} end
+Bottom0d     = BottomArray{0}
+BottomVector = BottomArray{1}
+BottomMatrix = BottomArray{2}
+# When two or more AbstractArrays have specialized broadcasting, and no `rule`
+# is defined to establish precedence, then we have a conflict
+abstract type ArrayConflict <: AbstractArray{Void,-1} end
+
+
+"""
+    result = Broadcast.Result{ContainerType}()
+    result = Broadcast.Result{ContainerType,ElType}(inds::Indices)
+
+Create an object that specifies the type and (optionally) indices
+of the result (output) of a broadcasting operation.
+
+Using a dedicated type for this information makes it possible to support
+variants of [`broadcast`](@ref) that accept `result` as an argument;
+it prevents an ambiguity of intent that would otherwise arise because
+both types and indices-tuples are among the supported *input*
+arguments to `broadcast`. For example, `parse.(Int, ("1", "2"))` is
+equivalent to `broadcast(parse, Int, ("1", "2"))`, and as a consequence
+it would would be ambiguous if result-type and output-indices information
+were passed as positional arguments to `broadcast`.
+
+You can extract `inds` with `indices(result)`.
+"""
+struct Result{ContainerType,ElType,I<:Union{Void,Indices}}
```

Comment:
IIUC, perhaps `ResultInfo` or something similar would better capture this type's purpose? And likewise `result` -> `resultinfo` / `r` -> `ri` for instances?